boards/nano33ble/src/test/linear_log_test.rs

```diff
@@ -0,0 +1,326 @@
+//! Tests the log storage interface in linear mode. For testing in circular mode, see
+//! log_test.rs.
+//!
+//! The testing framework creates a non-circular log storage interface in flash and performs a
+//! series of writes and syncs to ensure that the non-circular log properly denies overly-large
+//! writes once it is full. For testing all of the general capabilities of the log storage
+//! interface, see log_test.rs.
+//!
+//! To run the test, add the following line to the imix boot sequence:
+//! ```
+//!     test::linear_log_test::run(mux_alarm, dynamic_deferred_caller);
+//! ```
+//! and use the `USER` and `RESET` buttons to manually erase the log and reboot the imix,
+//! respectively.
+
+use capsules::log;
+use capsules::virtual_alarm::{MuxAlarm, VirtualMuxAlarm};
+use core::cell::Cell;
+use kernel::common::cells::{NumericCellExt, TakeCell};
+use kernel::common::dynamic_deferred_call::DynamicDeferredCall;
+use kernel::debug;
+use kernel::hil::flash;
+use kernel::hil::log::{LogRead, LogReadClient, LogWrite, LogWriteClient};
+use kernel::hil::time::{Alarm, AlarmClient};
+use kernel::static_init;
+use kernel::storage_volume;
+use kernel::ReturnCode;
+use nrf52840::{
+    nvmc::{NrfPage, Nvmc},
+    rtc::Rtc,
+};
+
+// Allocate 4 KiB volume for log storage (the nano33ble page size is 4 KiB).
+storage_volume!(LINEAR_TEST_LOG, 4);
+
+pub unsafe fn run(
+    mux_alarm: &'static MuxAlarm<'static, Rtc>,
+    deferred_caller: &'static DynamicDeferredCall,
+    flash_controller: &'static Nvmc,
+) {
+    // Set up flash controller.
+    flash_controller.configure_writeable();
+    flash_controller.configure_eraseable();
+    let pagebuffer = static_init!(NrfPage, NrfPage::default());
+
+    // Create actual log storage abstraction on top of flash.
+    let log = static_init!(
+        Log,
+        log::Log::new(
+            &LINEAR_TEST_LOG,
+            &flash_controller,
+            pagebuffer,
+            deferred_caller,
+            false
+        )
+    );
+    flash::HasClient::set_client(flash_controller, log);
+    log.initialize_callback_handle(
+        deferred_caller
+            .register(log)
+            .expect("no deferred call slot available for log storage"),
+    );
+
+    // Create and run test for log storage.
+    let test = static_init!(
+        LogTest<VirtualMuxAlarm<'static, Rtc>>,
+        LogTest::new(log, &mut BUFFER, VirtualMuxAlarm::new(mux_alarm), &TEST_OPS)
+    );
+    log.set_read_client(test);
+    log.set_append_client(test);
+    test.alarm.set_alarm_client(test);
+
+    test.run();
+}
+
+static TEST_OPS: [TestOp; 9] = [
+    TestOp::Read,
+    // Write to first page.
+    TestOp::Write(8),
+    TestOp::Write(300),
+    // Write to next page, too large to fit on first.
+    TestOp::Write(304),
+    // Write should fail, not enough space remaining.
```

Comment:
With the different page size, these comments no longer seem right.

Reply:
Updated! Thanks for bringing that to my attention.